### DISCUSSION TITLE
Running trials on multiple GPUs from a single node

### DISCUSSION THREAD

**Discussion by anons9** [optuna/optuna#3847] [Q&A]
I am looking for a way to run (distributed) trials on multiple GPUs (here: 4, using sqlite as local database) where I would not have to start a separate terminal for each call. What would you suggest is the best way to approach this?

I also stumbled upon this thread: https://github.com/optuna/optuna/issues/1365, however, I don't seem to understand as to why do we return GPU ID (as gpu_id) in objective function and how it is used later on.

Another question I would have is whether we should specify the study name explicitly when creating a study, as that is the practice when running from two different terminals.

Additionally (optional), if by any chance you have some familiarity, can distribution using the method in the link be combined with multiprocessing?

Thank you very much!

**Comment by nzw0301** [comment]
Suppose your individual objective requires only 1 GPU from your question (so if I'm wrong, please let me know that).

> I am looking for a way to run (distributed) trials on multiple GPUs (here: 4, using sqlite as local database) where I would not have to start a separate terminal for each call. What would you suggest is the best way to approach this?

I think one of the easiest way to use all GPUs is run a python code from 4 different processes like different terminals as explained in https://optuna.readthedocs.io/en/v2.1.0/tutorial/004_distributed.html. Note that we need to specify different GPU id in $\{0,1,2,3\}$ for each process. Note that as in the tutorial, SQLite is inappropriate DB for distributed optimisation.

> I also stumbled upon this thread: https://github.com/optuna/optuna/issues/1365, however, I don't seem to understand as to why do we return GPU ID (as gpu_id) in objective function and how it is used later on.

The code looks like a minimal example code; we don't need to return a GPU id as an objective value. For pytorch case, we might initialise device instance given GPU id.

> Another question I would have is whether we should specify the study name explicitly when creating a study, as that is the practice when running from two different terminals.

Yes. Otherwise, `study` instance cannot share the trials on the other processes.

**Comment by mdtanker** [comment]
Hopefully this is helpful to someone. Here is how I run an Optuna across multiple cores on the same machine using JobLib. Make sure to pip install tqdm_joblib if you want to get the progress bar. 

```python
import optuna
import psutil
import math
import joblib
from tqdm_joblib import tqdm_joblib

def optuna_max_cores(n_trials, optimize_study, study_name, study_storage, objective):
    """
    Set up optuna optimization in parallel splitting up the number of trials over all available cores.
    """
    # get available cores (UNIX and Windows)
    num_cores = len(psutil.Process().cpu_affinity())

    # set trials per job
    trials_per_job = math.ceil(n_trials/num_cores)

    # set number of jobs
    if n_trials >= num_cores:
        n_jobs = num_cores
    else:
        n_jobs = n_trials

    with tqdm_joblib(desc="Optimizing", total=n_trials) as progress_bar:
        joblib.Parallel(n_jobs=n_jobs)(
            joblib.delayed(optimize_study)(
                study_name,
                study_storage,
                objective,
                n_trials=trials_per_job,
                )
            for i in range(n_trials))

def optuna_parallel(
    study_name,
    study_storage,
    objective,
    n_trials=100,
    parallel=True,
):
    """
    Run optuna optimization in parallel. Pre-define the study, storage, and objective function and input them here.
    """
    # load study metadata from storage
    study = optuna.load_study(storage=study_storage, study_name=study_name)
    
    # set up parallel processing and run optimization
    if parallel is True:
        def optimize_study(study_name, storage, objective, n_trials):
            study = optuna.load_study(study_name=study_name, storage=storage)
            study.optimize(objective, n_trials=n_trials,)
            optuna_max_cores(n_trials, optimize_study, study_name, study_storage, objective)

    # run in normal, non-parallel mode
    elif parallel is False:
        study = optuna.load_study(
            study_name=study_name,
            storage=study_storage,
            )
        study.optimize(
            objective,
            n_trials=n_trials,
            show_progress_bar=True,
        )
    
    # reload the study
    study = optuna.load_study(
        study_name=study_name,
        storage=study_storage,
        )
    
    # get dataframe from study and sort by objective value
    study_df = study.trials_dataframe()
    
    return study, study_df

```

Implement it as below:
```python
# set name and storage for the optimization
study_name = "example"
storage = JournalStorage(JournalFileStorage(f"YOUR_LOCATION/{study_name}.log"))

n_trials=100

study = optuna.create_study(
    study_name=study_name,
    storage=storage,
    directions=["minimize", "minimize", "minimize"],
    sampler=optuna.samplers.TPESampler(n_startup_trials=int(n_trials/3)),
    load_if_exists=True,
    )

study, study_df = opti.optuna_parallel(
    study_name=study_name,
    study_storage=storage,
    objective = YOUR_OBJECTIVE_FUNCTION,
    n_trials=n_trials,
)
```
